sdk/identity/identity/src/credentials/onBehalfOfCredential.ts

```diff
@@ -0,0 +1,113 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+
+import { AccessToken, TokenCredential, GetTokenOptions } from "@azure/core-auth";
+import { TokenCredentialOptions } from "../client/identityClient";
+import { MsalOnBehalfOf } from "../msal/nodeFlows/msalOnBehalfOf";
+import { credentialLogger } from "../util/logging";
+import { trace } from "../util/tracing";
+import { CredentialPersistenceOptions } from "./credentialPersistenceOptions";
+
+const logger = credentialLogger("OnBehalfOfCredential");
+
+export interface OnBehalfOfCredentialOptions
+  extends TokenCredentialOptions,
+    CredentialPersistenceOptions {
+  userAssertion?: string;
+}
+
+/**
+ * Example usage:
+ *
+ * ```ts
+ * const credential = new OnBehalfOfCredential();
+ * const client = new SecretClient("https://key-vault-name.vault.azure.net", credential);
+ *
+ * // Must be called within a userAssertionContext(), so the following lines will fail:
+ * // await credential.getToken() will fail.
+ * // await client.getSecret("secret-name"); will fail.
+ *
+ * // This will work:
+ * await credential.userAssertionContext("user-assertion", async () => {
+ *   for await (const page of client.listPropertiesOfSecrets().byPage({ maxPageSize: 2 })) {
+ *     for (const secretProperties of page) {
+ *       if (secretProperties.enabled) {
+ *         const secret = await client.getSecret(secretProperties.name);
+ *         console.log("secret: ", secret);
+ *       }
+ *     }
+ *   }
+ * });
+ *
+ * // Any calls outside of `userAssertionContext` will fail.
+ * // Attempts to set multiple parallel userAssertionContexts will fail.
+ * ```
+ */
+export class OnBehalfOfCredential implements TokenCredential {
+  private options: OnBehalfOfCredentialOptions;
+  private contextLock: boolean = false;
+
+  /**
+   * Creates an instance of the OnBehalfOfCredential.
+   *
+   * @param tenantId - The Azure Active Directory tenant (directory) ID.
+   * @param clientId - The client (application) ID of an App Registration in the tenant.
+   * @param clientSecret - A client secret that was generated for the App Registration.
+   * @param options - Options for configuring the client which makes the authentication request.
+   */
+  constructor(
+    private tenantId: string,
+    private clientId: string,
+    private clientSecret: string,
+    options: OnBehalfOfCredentialOptions = {}
+  ) {
+    this.options = options;
+  }
+
+  async userAssertionContext(userAssertion: string, callback: () => Promise<void>): Promise<void> {
```

Comment:
I think we should change the name to reflect its function a bit better and also allow the consumer to return whatever they want from their callback to outer promise.

```suggestion
async withContext<Callback extends (...args:Args) => ReturnType<Callback>, Args extends unknown[]>(userAssertion: string, callback: Callback, ...args: Args): Promise<ReturnType<Callback>> {
```

I got a little extravagant here by permitting passing along arguments and supporting the callback both returning a promise or not, but I think it would feel great to be flexible, like:

```ts

// sync callback
const result  = await withContext("user", function() {
    return "foo";
});

// async callback, with argument
const result  = await withContext("user", async function() {
    const userData = await retrieveUserData(itemId);
    return userData;
}, requestedItemId);
```

Reply:
I don’t think I understand why we’re passing arguments to the callback, the idea is to only overwrite the getToken call inside of that scope, not to pass stuff back to the callback 🤔

Reply:
Callback being sync maybe never makes sense, so can change this to

```
withContext<Result, Callback extends (...args:Args) => Promise<Result>, Args extends unknown[]>(userAssertion: string, callback: Callback, ...args: Args): ReturnType<Callback>; 
```

Reply:
What do you think about this? https://github.com/Azure/azure-sdk-for-js/pull/16940/commits/c4b470b7a5cd69156fd908f798eba7925afe018f

Reply:
I particularly don’t like passing parameters through, since `withContext` does not enhance these parameters in any way. It seems misleading to me.

Reply:
Sure, I think the pattern is mostly for when you want to wrap an existing method, but that seems unlikely for this use case.

```
withContext<Result, Callback extends () => Promise<Result>>(userAssertion: string, callback: Callback): ReturnType<Callback>; 
```

That works, right?

Reply:
Yes! That works! I’ll clean this up and close the other PR.

Reply:
I also got confirmation from @schaabs that he and @bterlson were thinking of a similar idea yesterday :) so I think we just jumped ahead. I’ll consider the idea in this PR as the better solution amongst the alternatives!